File: app/src/main/java/org/ole/planet/myplanet/callback/ProgressBroadcastReceiver.java
package org.ole.planet.myplanet.callback;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import org.ole.planet.myplanet.model.Download;

import static org.ole.planet.myplanet.ui.dashboard.DashboardActivity.MESSAGE_PROGRESS;

public class ProgressBroadcastReceiver extends BroadcastReceiver {

    private OnProgressChange onProgressChange;

    public void setOnProgressChange(OnProgressChange onProgressChange) {
        this.onProgressChange = onProgressChange;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent.getAction().equals(MESSAGE_PROGRESS)) {
            Download download = intent.getParcelableExtra("download");
            if (onProgressChange != null) {
                onProgressChange.onProgressChange(String.format("Downloading file %d/%d KB\n%d%% Completed.", download.getCurrentFileSize(), download.getTotalFileSize(), download.getProgress()));
            }
        }
    }

    public interface OnProgressChange {
        void onProgressChange(String s);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/callback/ProgressBroadcastReceiver.kt
package org.ole.planet.myplanet.callback

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import org.ole.planet.myplanet.model.Download
import org.ole.planet.myplanet.ui.dashboard.DashboardActivity
import java.util.Locale

class ProgressBroadcastReceiver : BroadcastReceiver() {
    private var onProgressChange: OnProgressChange? = null

    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == DashboardActivity.MESSAGE_PROGRESS) {
            val download: Download? = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                intent.getParcelableExtra("download", Download::class.java)
            } else {
                @Suppress("DEPRECATION")
                intent.getParcelableExtra("download")
            }
            if (onProgressChange != null) {
                onProgressChange!!.onProgressChange(
                    String.format(
                        Locale.getDefault(),
                        "Downloading file %d/%d KB\n%d%% Completed.",
                        download!!.currentFileSize, download.totalFileSize, download.progress
                    )
                )
            }
        }
    }

    interface OnProgressChange {
        fun onProgressChange(s: String?)
    }
}
